### PR TITLE
Scheduler enhancements

### PR DESCRIPTION
**Pull request by kriegfrj**
This is a branch I have created to keep track of the scheduler enhancements I have been working on. Per @charlesmunger's suggestion, I'd like to keep these on a separate branch until they have been thoroughly tested, and it was getting difficult to maintain different PRs on different branches so I have consolidated them into one branch so that they can all be tested together. I think that @charlesmunger & @jongerrish have a good codebase to test it on, hopefully they get some headspace to put these changes through their paces.

For better understanding of this PR, please refer to issue #2149.

At this point in time, the tip of this branch contains the following enhancements:
- Nanosecond scheduler re-introduced (removed in #2119) and changes to `ShadowSystemClock` to avoid confusion around `setCurrentTimeMillis()` (ie, this branch also incorporates PR #2122).
- AsyncTask realism & background thread operation re-introduced (also removed in #2119).
- `Scheduler` class has been made fully thread-safe so that it shouldn't cause deadlocks when tasks are executed on non-main threads (this includes the changes in PR #2116, plus some more).
- `Scheduler` class also enforces that it can only be controlled (ie, paused or advanced) from the main thread (see #2149 for some rationale on this).

### CODE REVIEW

**Comment by karlicoss** [comment]
Hey @kriegfrj, great work! I ran your Robolectric branch against my codebase, and some tests got stuck. After some investigation, I figured out the minimal failing example:

```
import android.os.AsyncTask;
import android.os.Bundle;
import android.support.v4.app.Fragment;

import org.junit.Test;
import org.junit.runner.RunWith;
import org.robolectric.RobolectricGradleTestRunner;
import org.robolectric.shadows.support.v4.SupportFragmentController;

@RunWith(RobolectricGradleTestRunner.class)
public class FailingTest {

    @Test
    public void test() {
        final SupportFragmentController<TestFragment> fc = SupportFragmentController.of(new TestFragment());
        fc.create();
    }

    public static class TestFragment extends Fragment {

        @Override
        public void onCreate(Bundle savedInstanceState) {
            super.onCreate(savedInstanceState);
            new AsyncTask<Void, Void, Void>() {
                @Override
                protected Void doInBackground(Void... params) {
                    return null;
                }
            }.execute();
        }
    }
}
```

This is supposed to work normally, and finish properly (which does happen on stable Robolectric), however, the test stucks on your branch. Got any ideas why could this happen?

**Comment by kriegfrj** [comment]
Hi @karlicoss, thanks for the feedback! I suspect that this has something to do with the fact that the `FragmentController` (via `ActivityController`) is invoking the lifecycle methods through `runPaused()`. Hopefully i can have a closer look later tonight.

**Comment by kriegfrj** [comment]
Hi again @karlicoss, try again with the latest commit. It's a quick fix as I'm in the middle of rewriting that code anyway, but until the rewrite is ready this should fix your deadlock problem. 

Thanks a bunch for talking the time to find a reproducible test case, this was àn invaluable help!

**Comment by karlicoss** [comment]
That's cool @kriegfrj, no more hanging tests (I will leave the tests running in a loop overnight though to be sure). Aaand I found one more (potential) issue:

```
import android.os.AsyncTask;
import android.os.Handler;
import android.os.Looper;

import org.junit.Test;
import org.junit.runner.RunWith;
import org.robolectric.RobolectricGradleTestRunner;

import java.util.concurrent.atomic.AtomicBoolean;

import static org.junit.Assert.assertTrue;

@RunWith(RobolectricGradleTestRunner.class)
public class HandlerTest {

    @Test
    public void test() {
        final Handler mainHandler = new Handler(Looper.getMainLooper());
        AtomicBoolean called = new AtomicBoolean(false);
        new AsyncTask<Void, Void, Void>() {
            @Override
            protected Void doInBackground(Void... params) {
                mainHandler.post(
                        new Runnable() {
                            @Override
                            public void run() {
                                called.set(true);
                            }
                        }
                );
                return null;
            }
        }.execute();
        assertTrue(called.get());
    }
}
```

That test passes on stable robolectric, but fails on your branch, I think it is the `RuntimeEnvironment.isMainThread()` check in the `Sheduler.postDelayed` method.
But this could also be intended I guess since on a real device the code in the runnable will not actually get executed before the caller runnable finishes.

**Comment by kriegfrj** [comment]
Thanks, @karlicoss. Glad my firstpatch worked for you. The second issue is  definitely a bug.  I know what needs to be fixed here, I'll probably have a look some time after Christmas.

**Comment by kriegfrj** [comment]
Thanks, @karlicoss. Glad my firstpatch worked for you. The second issue is  definitely a bug.  I know what needs to be fixed here, I'll probably have a look some time after Christmas.

**Comment by karlicoss** [comment]
Hey @kriegfrj did you have a chance to look at it?

**Comment by kriegfrj** [comment]
Sorry @karlicoss, Christmas/New Year is a pretty busy time for us priests. :smile:

I have a patch in my local working copy which seems to be working. I've just been a bit reluctant to check it in yet because I might need to change it to implement the next phase. I'll try and push the latest changes tomorrow.

**Comment by kriegfrj** [comment]
Hi @karlicoss, I've just pushed a fix for the issue you raised. I also rebased it onto master (although master has since moved on again by the looks of things). This also fixes a couple of issues in and around #1306, although as I mentioned I anticipate needing to change this implementation once I implement the next pair of features (actually executing background loopers on background threads, and fixing sleep() behaviour). Let me know how it works for you and feel free to make suggestions.

**Comment by karlicoss** [comment]
So, I tested [this piece of code](https://github.com/robolectric/robolectric/pull/2166#issuecomment-166926094) again and it still fails. Tracing some of the execution steps:

1. `AsyncTask.execute` gets called on main thread
2. `AsyncTask.doInBackground` gets called on background thread
3. `mainHandler.post` calls `Scheduler.postDelayed`
4. current thread is still background, so the check `RuntimeEnvironment.isMainThread()` [here](https://github.com/robolectric/robolectric/pull/2166/files#diff-be64c725ac55c0a23eca7f3eac100229R192) doesn't pass and the runnable just gets into the queue without a change to execute.

**Comment by karlicoss** [comment]
Also, interesting thing: while debugging this test within my project, `doInBackground` is indeed executed on a separate thread (named `AsyncTask #1`). But, if I instead, create a test in the Robolectric project and run it, the task passes, but that's because `doInBackground` is executed on the main thread for some reason. I am investigating that.

**Comment by kriegfrj** [comment]
Thanks for the feedback. I know what's causing this and I should be able to fix it in the next couple days. The plan is to do something in ShadowWrangler so that every call from the main thread to an Android method is wrapped in a block()/unBlock() pair. That would fix your case because the unBlock() would get called at the end of execute(). This was my original plan in #1306, but I ran into some issues with ShadowLooper which I have since fixed.

**Comment by kriegfrj** [comment]
Hi @karlicoss, I think I've fixed that last problem you had with the post from background not working, courtesy of my fix for #1306. The latest commit has a regression test for it in `ShadowAsyncTaskTest`. If you get a chance, please try it out and let me know how you go.

**Comment by karlicoss** [comment]
Hmm, I am getting failures on robolectric tests, 

```
Tests run: 114, Failures: 11, Errors: 0, Skipped: 0, Time elapsed: 0.685 sec <<< FAILURE!
testNormalFlow[21](org.robolectric.shadows.ShadowAsyncTaskTest)  Time elapsed: 0.007 sec  <<< FAILURE!
java.lang.AssertionError: expected:<[onPostExecute c]> but was:<[]>
	at org.junit.Assert.fail(Assert.java:91)
	at org.junit.Assert.failNotEquals(Assert.java:645)
	at org.junit.Assert.assertEquals(Assert.java:126)
	at org.junit.Assert.assertEquals(Assert.java:145)
	at org.robolectric.util.Transcript.assertEventsSoFar(Transcript.java:24)
	at org.robolectric.shadows.ShadowAsyncTaskTest.testNormalFlow(ShadowAsyncTaskTest.java:119)

whenCancelled_beforeExecute_shouldCancelSuccessfully[21](org.robolectric.shadows.ShadowAsyncTaskTest)  Time elapsed: 0.004 sec  <<< FAILURE!
java.lang.AssertionError: expected:<[onCancelled null]> but was:<[]>
	at org.junit.Assert.fail(Assert.java:91)
	at org.junit.Assert.failNotEquals(Assert.java:645)
	at org.junit.Assert.assertEquals(Assert.java:126)
	at org.junit.Assert.assertEquals(Assert.java:145)
	at org.robolectric.util.Transcript.assertEventsSoFar(Transcript.java:24)
	at org.robolectric.shadows.ShadowAsyncTaskTest.whenCancelled_beforeExecute_shouldCancelSuccessfully(ShadowAsyncTaskTest.java:138)

whenCancelled_beforeBackgroundTasks_shouldCancelSuccessfully[21](org.robolectric.shadows.ShadowAsyncTaskTest)  Time elapsed: 0.004 sec  <<< FAILURE!
java.lang.AssertionError: expected:<[onCancelled null]> but was:<[]>
	at org.junit.Assert.fail(Assert.java:91)
	at org.junit.Assert.failNotEquals(Assert.java:645)
	at org.junit.Assert.assertEquals(Assert.java:126)
	at org.junit.Assert.assertEquals(Assert.java:145)
	at org.robolectric.util.Transcript.assertEventsSoFar(Transcript.java:24)
	at org.robolectric.shadows.ShadowAsyncTaskTest.whenCancelled_beforeBackgroundTasks_shouldCancelSuccessfully(ShadowAsyncTaskTest.java:157)

whenCancelled_beforeFinishCallback_shouldCallOnCancelled[21](org.robolectric.shadows.ShadowAsyncTaskTest)  Time elapsed: 0.003 sec  <<< FAILURE!
java.lang.AssertionError: expected:<[onCancelled c]> but was:<[]>
	at org.junit.Assert.fail(Assert.java:91)
	at org.junit.Assert.failNotEquals(Assert.java:645)
	at org.junit.Assert.assertEquals(Assert.java:126)
	at org.junit.Assert.assertEquals(Assert.java:145)
	at org.robolectric.util.Transcript.assertEventsSoFar(Transcript.java:24)
	at org.robolectric.shadows.ShadowAsyncTaskTest.whenCancelled_beforeFinishCallback_shouldCallOnCancelled(ShadowAsyncTaskTest.java:176)

whenCancelled_afterFinishCallback_shouldCallOnPostExecute[21](org.robolectric.shadows.ShadowAsyncTaskTest)  Time elapsed: 0.004 sec  <<< FAILURE!
java.lang.AssertionError: expected:<[onPostExecute c]> but was:<[]>
	at org.junit.Assert.fail(Assert.java:91)
	at org.junit.Assert.failNotEquals(Assert.java:645)
	at org.junit.Assert.assertEquals(Assert.java:126)
	at org.junit.Assert.assertEquals(Assert.java:145)
	at org.robolectric.util.Transcript.assertEventsSoFar(Transcript.java:24)
	at org.robolectric.shadows.ShadowAsyncTaskTest.whenCancelled_afterFinishCallback_shouldCallOnPostExecute(ShadowAsyncTaskTest.java:195)

progressUpdates_shouldBeQueued_untilBackgroundThreadFinishes[21](org.robolectric.shadows.ShadowAsyncTaskTest)  Time elapsed: 0.003 sec  <<< FAILURE!
java.lang.AssertionError: expected:<[onProgressUpdate 33%, onProgressUpdate 66%, onProgressUpdate 99%, onPostExecute done]> but was:<[]>
	at org.junit.Assert.fail(Assert.java:91)
	at org.junit.Assert.failNotEquals(Assert.java:645)
	at org.junit.Assert.assertEquals(Assert.java:126)
	at org.junit.Assert.assertEquals(Assert.java:145)
	at org.robolectric.util.Transcript.assertEventsSoFar(Transcript.java:24)
	at org.robolectric.shadows.ShadowAsyncTaskTest.progressUpdates_shouldBeQueued_untilBackgroundThreadFinishes(ShadowAsyncTaskTest.java:222)

progressUpdates_shouldStopBeingQueued_onceAborted[21](org.robolectric.shadows.ShadowAsyncTaskTest)  Time elapsed: 0.005 sec  <<< FAILURE!
java.lang.AssertionError: expected:<[onProgressUpdate 33%]> but was:<[]>
	at org.junit.Assert.fail(Assert.java:91)
	at org.junit.Assert.failNotEquals(Assert.java:645)
	at org.junit.Assert.assertEquals(Assert.java:126)
	at org.junit.Assert.assertEquals(Assert.java:145)
	at org.robolectric.util.Transcript.assertEventsSoFar(Transcript.java:24)
	at org.robolectric.shadows.ShadowAsyncTaskTest.progressUpdates_shouldStopBeingQueued_onceAborted(ShadowAsyncTaskTest.java:250)

shouldGetStatusForAsyncTask[21](org.robolectric.shadows.ShadowAsyncTaskTest)  Time elapsed: 0.009 sec  <<< FAILURE!
org.junit.ComparisonFailure: [after] expected:<[FINISHED]> but was:<[RUNNING]>
	at sun.reflect.NativeConstructorAccessorImpl.newInstance0(Native Method)
	at sun.reflect.NativeConstructorAccessorImpl.newInstance(NativeConstructorAccessorImpl.java:62)
	at sun.reflect.DelegatingConstructorAccessorImpl.newInstance(DelegatingConstructorAccessorImpl.java:45)
	at org.robolectric.shadows.ShadowAsyncTaskTest.shouldGetStatusForAsyncTask(ShadowAsyncTaskTest.java:286)

onPostExecute_doesNotSwallowExceptions[21](org.robolectric.shadows.ShadowAsyncTaskTest)  Time elapsed: 0.004 sec  <<< FAILURE!
java.lang.AssertionError: Expected RuntimeException to be thrown
	at org.robolectric.shadows.ShadowAsyncTaskTest.onPostExecute_doesNotSwallowExceptions(ShadowAsyncTaskTest.java:310)

executeOnExecutor_withImmediateExecutor_doesntBlock[21](org.robolectric.shadows.ShadowAsyncTaskTest)  Time elapsed: 0.004 sec  <<< FAILURE!
org.junit.ComparisonFailure: expected:<[FINISHED]> but was:<[RUNNING]>
	at sun.reflect.NativeConstructorAccessorImpl.newInstance0(Native Method)
	at sun.reflect.NativeConstructorAccessorImpl.newInstance(NativeConstructorAccessorImpl.java:62)
	at sun.reflect.DelegatingConstructorAccessorImpl.newInstance(DelegatingConstructorAccessorImpl.java:45)
	at org.robolectric.shadows.ShadowAsyncTaskTest.executeOnExecutor_withImmediateExecutor_doesntBlock(ShadowAsyncTaskTest.java:330)

executeOnExecutor_withParallelExecutor_andMultipleTasks[21](org.robolectric.shadows.ShadowAsyncTaskTest)  Time elapsed: 0.006 sec  <<< FAILURE!
java.lang.AssertionError: expected:<[onPostExecute c, onPostExecute c, onPostExecute c, onPostExecute c]> but was:<[]>
	at org.junit.Assert.fail(Assert.java:91)
	at org.junit.Assert.failNotEquals(Assert.java:645)
	at org.junit.Assert.assertEquals(Assert.java:126)
	at org.junit.Assert.assertEquals(Assert.java:145)
	at org.robolectric.util.Transcript.assertEventsSoFar(Transcript.java:24)
	at org.robolectric.shadows.ShadowAsyncTaskTest.executeOnExecutor_withParallelExecutor_andMultipleTasks(ShadowAsyncTaskTest.java:351)
```

Maybe I messed up somewhere, let's see what the CI says.

**Comment by kriegfrj** [comment]
For some reason Travis seems not to have triggered a build for the master repository on my latest commit to this branch (or at least, I haven't been able to find a record of it). My own Travis installation did trigger a build when I pushed it and worked; see https://travis-ci.org/kriegfrj/robolectric/builds/104104034

Sorry if I'm staying the obvious, but have you done a clean install (I did notice that all the failures were for API 21??)

**Comment by karlicoss** [comment]
Yeah, only api 21, that's true. I think I did, but anyway gonna try one more time :)

**Comment by karlicoss** [comment]
Ok, so one of my tests broke which relied on `ShadowSystemClock.currentTimeMillis` being non-zero by default, but that's okay, I guess I shouldn't have relied of that in the first place. Maybe worth mentioning in release notes.

**Comment by karlicoss** [comment]
And a lot of tests failed with something like

```
...
at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:50)
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:47)
	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
	at org.robolectric.RobolectricTestRunner$HelperTestRunner$1.evaluate(RobolectricTestRunner.java:548)
	at org.junit.internal.runners.statements.RunBefores.evaluate(RunBefores.java:26)
	at org.robolectric.RobolectricTestRunner$2.evaluate(RobolectricTestRunner.java:255)
	at org.robolectric.RobolectricTestRunner.runChild(RobolectricTestRunner.java:185)
	at org.robolectric.RobolectricTestRunner.runChild(RobolectricTestRunner.java:54)
	at org.junit.runners.ParentRunner$3.run(ParentRunner.java:290)
	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:71)
	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:288)
	at org.junit.runners.ParentRunner.access$000(ParentRunner.java:58)
	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:268)
	at org.robolectric.RobolectricTestRunner$1.evaluate(RobolectricTestRunner.java:151)
	at org.junit.runners.ParentRunner.run(ParentRunner.java:363)
	at org.gradle.api.internal.tasks.testing.junit.JUnitTestClassExecuter.runTestClass(JUnitTestClassExecuter.java:105)
	at org.gradle.api.internal.tasks.testing.junit.JUnitTestClassExecuter.execute(JUnitTestClassExecuter.java:56)
	at org.gradle.api.internal.tasks.testing.junit.JUnitTestClassProcessor.processTestClass(JUnitTestClassProcessor.java:64)
	at org.gradle.api.internal.tasks.testing.SuiteTestClassProcessor.processTestClass(SuiteTestClassProcessor.java:50)
	at org.gradle.messaging.dispatch.ReflectionDispatch.dispatch(ReflectionDispatch.java:35)
	at org.gradle.messaging.dispatch.ReflectionDispatch.dispatch(ReflectionDispatch.java:24)
	at org.gradle.messaging.dispatch.ContextClassLoaderDispatch.dispatch(ContextClassLoaderDispatch.java:32)
	at org.gradle.messaging.dispatch.ProxyDispatchAdapter$DispatchingInvocationHandler.invoke(ProxyDispatchAdapter.java:93)
	at com.sun.proxy.$Proxy2.processTestClass(Unknown Source)
	at org.gradle.api.internal.tasks.testing.worker.TestWorker.processTestClass(TestWorker.java:106)
	at org.gradle.messaging.dispatch.ReflectionDispatch.dispatch(ReflectionDispatch.java:35)
	at org.gradle.messaging.dispatch.ReflectionDispatch.dispatch(ReflectionDispatch.java:24)
	at org.gradle.messaging.remote.internal.hub.MessageHub$Handler.run(MessageHub.java:360)
	at org.gradle.internal.concurrent.ExecutorPolicy$CatchAndRecordFailures.onExecute(ExecutorPolicy.java:54)
	at org.gradle.internal.concurrent.StoppableExecutorImpl$1.run(StoppableExecutorImpl.java:40)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
	at java.lang.Thread.run(Thread.java:745)
Caused by: java.lang.ClassNotFoundException: javax.microedition.khronos.opengles.GL
	at org.robolectric.internal.bytecode.InstrumentingClassLoader.loadClass(InstrumentingClassLoader.java:87)
	at java.lang.Class.getDeclaredMethods0(Native Method)
	at java.lang.Class.privateGetDeclaredMethods(Class.java:2701)
	at java.lang.Class.getDeclaredMethod(Class.java:2128)
	at org.robolectric.internal.bytecode.ShadowWrangler.getRealMethodPlan(ShadowWrangler.java:123)
	at org.robolectric.internal.bytecode.ShadowWrangler.calculatePlan(ShadowWrangler.java:165)
	at org.robolectric.internal.bytecode.ShadowWrangler.methodInvoked(ShadowWrangler.java:114)
	at org.robolectric.internal.bytecode.RobolectricInternals.methodInvoked(RobolectricInternals.java:23)
	at android.graphics.Canvas.<init>(Canvas.java)
	at android.view.Surface$CompatibleCanvas.<init>(Surface.java)
	at android.view.Surface$CompatibleCanvas.<init>(Surface.java)
	at android.view.Surface.$$robo$$__constructor__(Surface.java:76)
	at sun.reflect.GeneratedMethodAccessor3129.invoke(Unknown Source)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:497)
	at org.robolectric.internal.bytecode.ShadowWrangler$RealMethodPlan.run(ShadowWrangler.java:458)
	at android.view.Surface.<init>(Surface.java)
	at android.view.ViewRootImpl.$$robo$$__constructor__(ViewRootImpl.java:243)
	at sun.reflect.GeneratedMethodAccessor3128.invoke(Unknown Source)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:497)
	at org.robolectric.internal.bytecode.ShadowWrangler$RealMethodPlan.run(ShadowWrangler.java:458)
	at android.view.ViewRootImpl.<init>(ViewRootImpl.java)
	at android.view.WindowManagerGlobal.$$robo$$addView(WindowManagerGlobal.java:239)
	at sun.reflect.GeneratedMethodAccessor3123.invoke(Unknown Source)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:497)
	at org.robolectric.internal.bytecode.ShadowWrangler$RealMethodPlan.run(ShadowWrangler.java:458)
	at android.view.WindowManagerGlobal.addView(WindowManagerGlobal.java)
	at android.view.WindowManagerImpl.$$robo$$addView(WindowManagerImpl.java:69)
	at sun.reflect.GeneratedMethodAccessor3122.invoke(Unknown Source)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:497)
	at org.robolectric.util.ReflectionHelpers.callInstanceMethod(ReflectionHelpers.java:195)
	at org.robolectric.internal.Shadow.directlyOn(Shadow.java:44)
	at org.robolectric.internal.Shadow.directlyOn(Shadow.java:36)
	at org.robolectric.shadows.ShadowWindowManagerImpl.addView(ShadowWindowManagerImpl.java:29)
	at sun.reflect.GeneratedMethodAccessor3121.invoke(Unknown Source)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:497)
	at org.robolectric.internal.bytecode.ShadowWrangler$ShadowMethodPlan.run(ShadowWrangler.java:488)
	at android.view.WindowManagerImpl.addView(WindowManagerImpl.java)
	at android.app.Activity.$$robo$$makeVisible(Activity.java:4100)
	at sun.reflect.GeneratedMethodAccessor3119.invoke(Unknown Source)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:497)
	at org.robolectric.internal.bytecode.ShadowWrangler$RealMethodPlan.run(ShadowWrangler.java:458)
	at android.app.Activity.makeVisible(Activity.java)
	at sun.reflect.GeneratedMethodAccessor3118.invoke(Unknown Source)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:497)
	at org.robolectric.util.ReflectionHelpers$3.run(ReflectionHelpers.java:162)
	at org.robolectric.util.ReflectionHelpers.traverseClassHierarchy(ReflectionHelpers.java:305)
	at org.robolectric.util.ReflectionHelpers.callInstanceMethod(ReflectionHelpers.java:156)
	at org.robolectric.util.ActivityController$2.run(ActivityController.java:182)
	at org.robolectric.shadows.CoreShadowsAdapter$2.runPaused(CoreShadowsAdapter.java:46)
	at org.robolectric.util.ActivityController.visible(ActivityController.java:178)
	at com.yandex.mail.tools.FragmentController.visible(FragmentController.java:95)
	at com.yandex.mail.compose.ComposeFragmentTest.prepareFragment(ComposeFragmentTest.java:143)
	at com.yandex.mail.compose.ComposeFragmentTest.compose_loadsSignature(ComposeFragmentTest.java:435)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:497)
	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:50)
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:47)
	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
	at org.robolectric.RobolectricTestRunner$HelperTestRunner$1.evaluate(RobolectricTestRunner.java:548)
	at org.junit.internal.runners.statements.RunBefores.evaluate(RunBefores.java:26)
	at org.robolectric.RobolectricTestRunner$2.evaluate(RobolectricTestRunner.java:255)
	at org.robolectric.RobolectricTestRunner.runChild(RobolectricTestRunner.java:185)
	at org.robolectric.RobolectricTestRunner.runChild(RobolectricTestRunner.java:54)
	at org.junit.runners.ParentRunner$3.run(ParentRunner.java:290)
	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:71)
	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:288)
	at org.junit.runners.ParentRunner.access$000(ParentRunner.java:58)
	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:268)
	at org.robolectric.RobolectricTestRunner$1.evaluate(RobolectricTestRunner.java:151)
	at org.junit.runners.ParentRunner.run(ParentRunner.java:363)
	at org.gradle.api.internal.tasks.testing.junit.JUnitTestClassExecuter.runTestClass(JUnitTestClassExecuter.java:105)
	at org.gradle.api.internal.tasks.testing.junit.JUnitTestClassExecuter.execute(JUnitTestClassExecuter.java:56)
	at org.gradle.api.internal.tasks.testing.junit.JUnitTestClassProcessor.processTestClass(JUnitTestClassProcessor.java:64)
	at org.gradle.api.internal.tasks.testing.SuiteTestClassProcessor.processTestClass(SuiteTestClassProcessor.java:50)
	at sun.reflect.GeneratedMethodAccessor756.invoke(Unknown Source)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:497)
	at org.gradle.messaging.dispatch.ReflectionDispatch.dispatch(ReflectionDispatch.java:35)
	at org.gradle.messaging.dispatch.ReflectionDispatch.dispatch(ReflectionDispatch.java:24)
	at org.gradle.messaging.dispatch.ContextClassLoaderDispatch.dispatch(ContextClassLoaderDispatch.java:32)
	at org.gradle.messaging.dispatch.ProxyDispatchAdapter$DispatchingInvocationHandler.invoke(ProxyDispatchAdapter.java:93)
	at com.sun.proxy.$Proxy2.processTestClass(Unknown Source)
	at org.gradle.api.internal.tasks.testing.worker.TestWorker.processTestClass(TestWorker.java:106)
	at sun.reflect.GeneratedMethodAccessor755.invoke(Unknown Source)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:497)
	... 8 more
```

, gonna check, might be my fault.

**Comment by karlicoss** [comment]
@kriegfrj I guess the stacktrace above was caused by your last commit, there is `getRealMethodPlan` invokation. I hope you could figure out what could be wrong from the stacktrace, but tell if you need more insight.

**Comment by kriegfrj** [comment]
Hi @karlicoss, thank you so much for putting in the effort to put these changes through their paces.

I'll try and get to the bottom of it, but if you've got a little bit more information (better still a reproducible test case) that would help greatly. I'm particularly interested in knowing if it is the same class triggering the NotFoundException in each case.

**Comment by karlicoss** [comment]
Oh, it actually turned out to be easy to reproduce: just try running

```
AlertDialog alertDialog = new AlertDialog.Builder(RuntimeEnvironment.application).create();
alertDialog.show();
```

weird though that Robolectric does have tests for AlertDialog, and doesn't crash on them

**Comment by karlicoss** [comment]
And another way to reproduce 

```
@Test
public void fefwef() {
    FragmentController<AlalaFragment> controller = FragmentController.of(new AlalaFragment());
    controller.create().start().resume().visible();
}

public static class AlalaFragment extends Fragment {
}
```

**Comment by karlicoss** [comment]
Also, that's interesting: https://github.com/robolectric/robolectric/issues/1932

**Comment by karlicoss** [comment]
Tried running once again against your previous commit (https://github.com/kriegfrj/robolectric/commit/7b6290803fd340bdce700e0f940bc030ee7cf7f9), and the snippets I posted above work fine, so it definitely something in the last commit :) Might be possible, that you just accidentally raised the issue that was already present https://github.com/robolectric/robolectric/issues/1932

**Comment by karlicoss** [comment]
Aaand I managed to reproduce it even on master Robolectric:

```
Class<Canvas> canvas = Canvas.class;
canvas.getDeclaredMethods();
```

So in you last commit you aded some `getDeclaredMethods` calls, and Robolectric tried to load this `GL` class. So it's not really your fault, and I guess has to be resolved in a separate PR.

**Comment by kriegfrj** [comment]
Thanks for all your detective work, @karlicoss!

I'm guessing the reason that my change triggered #1932 is because effectively the latest commit has changed the `InstrumentingClassLoader` behaviour from lazy to eager - whereas before it would only try to load classes from method signatures when those methods were invoked (and hence never tries at all unless the method is actually invoked at some point), now it tries to load them all when the class is first loaded (regardless of whether or not the class is loaded). It is possible that I could change this back to lazy again by deferring the class parameter & method lookup until the first time `RealMethodPlan.run()` is invoked; however I'm thinking that for testing purposes eager loading might actually be a good thing by helping to catch these loading problems early?

**Comment by karlicoss** [comment]
@kriegfrj yeah, that's what I was thinking, like I said here https://github.com/robolectric/robolectric/issues/1932#issuecomment-174333205, looks like it doesn't fail in Robolectric tests, because `android-all` artifact ends up in "compile classpath" (not sure whether the terminology is correct :) ) and loads via system classloaded, whereas when we use Robolectric in an android project, `android-all` has provided scope, and can only be loaded via `InstrumentingClassLoader` (but does not since the whole `javax.` package is whitelisted).

So if we had some integration tests which were external to Robolectric, it could be noticed earlier. Perhaps even not only eager loading problems, after all class loading is weird and complicated, who knows :)

**Comment by karlicoss** [comment]
@kriegfrj hey any progress?

**Comment by kriegfrj** [comment]
you mean on the #1932 related issue or on the next scheduling enhancement? I have no plans to look at #1932 for now, but i can make a change to the previous feature so that it doesn't make #1932 manifest more often than it does in master. I will look at that tomorrow. The next feature shouldn't be too far off either.

**Comment by karlicoss** [comment]
@kriegfrj yeah, I meant the change to lazy behaviour. I mean, if it gets merged, it will definitely break a lot of real tests. Although, some progress seems to be going here https://github.com/robolectric/robolectric/issues/1864, which is basically the same issue, so maybe by the time you finish your PR the issue will be gone.

**Comment by kriegfrj** [comment]
Hey @karlicoss, I had a bit of a look at the lazy loading issue, and I've decided not to fix it just now for a few reasons:

1. It proved a bit harder than I thought (in fact I'm not entirely convinced it was my changes that caused it).
2. I may need to change the implementation slightly anyway to accommodate the last two features, which might make the problem obsolete,
3. I'm hoping (as you suggested) that by the time I  finish this PR the root cause will be fixed anyway.

In short, I'm engaging in some lazy evaluation of my own. :smile: Hoping to have the next feature implemented in the next day or so. Thanks for your patience.

**Comment by kriegfrj** [comment]
Just touching base - been super busy in "real life" in the last couple of months, which is really irritating because I felt I was not far away from having this finished. Hopefully soon the skies will clear...

**Comment by IcyMidnight** [comment]
Any chance of getting this revived? Running into deadlocks in one of my projects related to Scheduler.getCurrentTime()

**Comment by kriegfrj** [comment]
I am really sorry, but I personally don't have the time to fix this at the moment, except in the unlikely event that the community might be able to sponsor me. The annoying thing is that it was nearly working, but by now I'll have a huge merge/rebase conflict to fix before I could even get back to finishing the last bit. I apologise to the community.

**Comment by IcyMidnight** [comment]
I understand. What about #2116? Could we get that one in to at least remove one deadlock? I was considering putting that change or something very similar forward on its own.

**Comment by kriegfrj** [comment]
You could try and incorporate the fix for #2116 in isolation of the other bits of the package - it might solve your immediate deadlock problem. If you wanted you could probably lift the changes to Scheduler from the old PR to achieve this; give it a try and if it works perhaps try and submit a new PR with just that change.

**Comment by utzcoz** [comment]
@kriegfrj Thanks for your contribution. This PR is old and has some conflicts with current master branch. I will close this PR. If you want to continue this PR, feel free to reopen this PR, rebase it on the latest master branch and squash commits to the one commit for reviewing. Thanks.
